### PR TITLE
Manage membership state, deleting users implemented.

### PR DESCRIPTION
**Pull request by vranki**
Basic implementation. UI may need more polishing but that can be done in another task.

### CODE REVIEW

File: www/views.py [hunk 1]
```diff
@@ -458,3 +458,45 @@ def createuser(request):
...
+                    _(
+                        "User not member or requesting delete"
+                    ),
+                )
+        return userdetails(request, id)
+
+    custominvoices = CustomInvoice.objects.filter(user=user)
+
+    for invoice in custominvoices:
+        invoice.delete()
```
Comment (tswfi) [inline www/views.py:479]:
```
CustomInvoice.objects.filter(user=user).delete()
```
Would be faster and in some cases does not require loading all the objects into memory before deletion

File: www/templates/www/user.html [hunk 2]
```diff
@@ -5,12 +5,14 @@
...
 <div class="container">
   <h1>
-    {% trans 'User data for' %} {{ userdetails.first_name }} {{ userdetails.last_name }}
+    {% trans 'User data for' %} {{ userdetails }}
   </h1>
 
   {% if user.is_staff %}
     <p>
       <a href="{% url 'updateuser' userdetails.id %}">Update user's data</a> |
+      <a href="{% url 'users/delete' userdetails.id 0 %}">D</a>elete user - keep account |
```
Comment (tswfi) [inline www/templates/www/user.html:14]: I'm guessing the link is for only the D letter for fat finger protection?

File: users/models.py [hunk 3]
```diff
@@ -93,7 +93,8 @@ class Meta:
 
     email = models.EmailField(
         unique=True,
-        blank=False,
+        blank=True,
+        null=True,
```
Comment (tswfi) [inline users/models.py:97]: unique=true and null=true?

Can we figure out a better way for this? for example setting the email address to devnull<originalid>@tampere.hacklab.fi ?

Or at least I would like to see a new objectmanager that can be used to always only fetch "active" users. Something like CustomUser.activeUsers.all(). (Or even in such a way that objects always returns only active and have another allObjects manager that returns also the "deleted" users.

File: www/views.py [hunk 4]
```diff
@@ -458,3 +458,45 @@ def createuser(request):
...
+
+    subscriptions = ServiceSubscription.objects.filter(user=user)
```
Comment (tswfi) [inline www/views.py:481]: see above :)

File: www/views.py [hunk 5]
```diff
@@ -458,3 +458,45 @@ def createuser(request):
...
+    for subscription in subscriptions:
+        subscription.delete()
+
```
Comment (tswfi) [inline www/views.py:484]: What about the banktransaction "sender" information?
```
for transaction in BankTransaction.objects.filter(user=user):
  transaction.sender="anonymized"
  transaction.save()
```

(untested)

File: www/views.py (hunk 1 above)
Comment (vranki) [reply www/views.py:479]: Done's

File: users/models.py (hunk 3 above)
Comment (vranki) [reply users/models.py:97]: I don't like fake empty values such as devnull suggested. Null & blank is better if the value can be blank as it now can be. We just need to make sure that it's not empty on registration and can become empty only on resignation.

File: www/views.py (hunk 5 above)
Comment (vranki) [reply www/views.py:484]: As discussed, sender in bank data is not personal info and isn't necessarily same as the user was.

Comment (tswfi) [reply www/views.py:482]: it is personal info

https://tietosuoja.fi/mika-on-henkilotieto#:~:text=Henkil%C3%B6tietoja%20ovat%20kaikki%20tiedot%2C%20jotka,toiseen%20tietoon%2C%20joka%20mahdollistaa%20tunnistamisen.

it doesn't matter where that info exists in the db it is still personal information.

we have no legal basis on keeping that data.

File: www/templates/www/user.html (hunk 2 above)
Comment (vranki) [reply www/templates/www/user.html:14]: Yep

**Comment by olmari** [comment]
For closure of earlier thinking process: banking data storage/archival is usually done elsewhere than on membership manager, thus the "first option" should be preferred in this case.

**Comment by tswfi** [comment]
I'm closing this. Its a bit old now.. Please reopen if you plan to implement this further

**Pull request closed without merging**